src/Rotation.cpp

```diff
@@ -0,0 +1,348 @@
+#include "Rotation.h"
+
+#include <algorithm>
+#include <exception>
+
+#include <Eigen/Geometry>
+
+namespace ale {
+
+///////////////////////////////////////////////////////////////////////////////
+// Helper Functions
+///////////////////////////////////////////////////////////////////////////////
+
+  Eigen::Vector3d axis(int axisIndex) {
+    switch (axisIndex) {
+      case 0:
+        return Eigen::Vector3d::UnitX();
+        break;
+      case 1:
+        return Eigen::Vector3d::UnitY();
+        break;
+      case 2:
+        return Eigen::Vector3d::UnitZ();
+        break;
+      default:
+        throw std::invalid_argument("Axis index must be 0, 1, or 2.");
+    }
+  }
+
+
+  // This is actually the transpose of the skew AV matrix because we define AV
+  // as the AV from the destination to the source. This matches how NAIF
+  // defines AV.
+  Eigen::Quaterniond::Matrix3 avSkewMatrix(
+        const std::vector<double>& av
+  ) {
+    if (av.size() != 3) {
+      throw std::invalid_argument("Angular velocity vector to rotate is the wrong size.");
+    }
+    Eigen::Quaterniond::Matrix3 avMat;
+    avMat <<  0.0,    av[2], -av[1],
+             -av[2],  0.0,    av[0],
+              av[1], -av[0],  0.0;
+    return avMat;
+  }
+
+  ///////////////////////////////////////////////////////////////////////////////
+  // Rotation Impl class
+  ///////////////////////////////////////////////////////////////////////////////
+
+  class Rotation::Impl {
+    public:
+      Impl() : quat(Eigen::Quaterniond::Identity()) { }
+
+
+      Impl(double w, double x, double y, double z) : quat(w, x, y, z) { }
+
+
+      Impl(const std::vector<double>& matrix) {
+        if (matrix.size() != 9) {
+          throw std::invalid_argument("Rotation matrix must be 3 by 3.");
+        }
+        quat = Eigen::Quaterniond(Eigen::Quaterniond::Matrix3(matrix.data()));
+      }
+
+
+      Impl(const std::vector<double>& angles, const std::vector<int>& axes) {
+        if (angles.empty() || axes.empty()) {
+          throw std::invalid_argument("Angles and axes must be non-empty.");
+        }
+        if (angles.size() != axes.size()) {
+          throw std::invalid_argument("Number of angles and axes must be equal.");
+        }
+        quat = Eigen::Quaterniond::Identity();
+
+        for (size_t i = 0; i < angles.size(); i++) {
+          quat *= Eigen::Quaterniond(Eigen::AngleAxisd(angles[i], axis(axes[i])));
+        }
+      }
+
+
+      Impl(const std::vector<double>& axis, double theta) {
```

Comment:
Axis and angle seem swapped compared to the above contractor.

Reply:
These are largely two separate things. The first angle + axes constructor takes Euler angles with explicit X, Y, Z axes. This constructor is for a rotations around an arbitrary axis (i.e. a 120 degree rotation around the axis 1, 1, 1).

I can swap these arguments so that they are more similar, but I don't want users to assume that the axes argument in the first constructor and axis argument in the second constructor are the same.

Reply:
That seems more of a documentation/naming issue, swapping them isn't going to do much to make it seem they are different things.